src/rt-app_taskgroups.c

```diff
@@ -0,0 +1,365 @@
+#include <stdlib.h>
+#include <string.h>
+#include <mntent.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <stdio.h>
+#include <errno.h>
+
+#include "config.h"
+#include "rt-app_utils.h"
+#include "rt-app_taskgroups.h"
+
+typedef struct _taskgroup_ctrl_t
+{
+	char mount_point[FILENAME_MAX];
+} taskgroup_ctrl_t;
+
+static taskgroup_ctrl_t tg_ctrl;
+
+static int cgroup_check_cpu_controller(void)
+{
+	int dummy[3], ret = 0;
+	FILE *cgroups;
+	char buf[512];
+
+	cgroups = fopen("/proc/cgroups", "re");
+
+	if (!cgroups) {
+		perror("fopen");
+		goto err;
+	}
+
+	/* Ignore the first line as it contains the header. */
+	if (!fgets(buf, sizeof(buf), cgroups)) {
+		perror("fgets");
+		goto err;
+	}
+
+	while (!feof(cgroups)) {
+		if (fscanf(cgroups, "%s %d %d %d", buf, &dummy[0], &dummy[1],
+			   &dummy[2]) < 4) {
+			perror("fscanf");
+			goto err;
+		}
+
+		if (!strcmp(buf, "cpu"))
+			goto done;
+	}
+err:
+	ret = -1;
+done:
+	if (cgroups)
+		fclose(cgroups);
+
+	return ret;
+}
+
+static int cgroup_get_cpu_controller_mount_point(void)
+{
+	struct mntent *ent;
+	FILE *mounts;
+	int ret = -1;
+
+	mounts = fopen("/proc/mounts", "re");
+
+	if (!mounts) {
+		perror("fopen");
+		return ret;
+	}
+
+	while (ent = getmntent(mounts)) {
+		if (strcmp(ent->mnt_type, "cgroup"))
+			continue;
+
+		if (!hasmntopt(ent, "cpu"))
+			continue;
+
+		strcpy(tg_ctrl.mount_point, ent->mnt_dir);
+
+		log_debug("cgroup cpu controller mountpoint [%s] found", ent->mnt_dir);
+		ret = 0;
+		break;
+	}
+
+	fclose(mounts);
+	return ret;
+}
+
+static int cgroup_attach_task(char *name)
+{
+	char *path;
+	FILE *tasks;
+	int ret = -1;
+
+	path = malloc(FILENAME_MAX);
+	if (!path) {
+		perror("malloc");
+		goto error;
+	}
+
+	sprintf(path, "%s", tg_ctrl.mount_point);
+
+	if (!name)
+		goto next;
+
+	if (strlen(name) > FILENAME_MAX - strlen(path) - 1) {
+		log_error("cgroup [%s] no space to create cgroup filename", name);
+		goto error;
+	}
+
+	strcat(path, name);
+next:
+	if (strlen("/tasks") > FILENAME_MAX - strlen(path) - 1) {
+		log_error("cgroup [%s] no space to create /tasks filename", name);
+		goto error;
+	}
+
+	strcat(path, "/tasks");
+
+	tasks = fopen(path, "we");
+	if (!tasks) {
+		perror("fopen");
+		goto error;
+	}
+
+	if (fprintf(tasks, "%ld", syscall(SYS_gettid)) < 0) {
```

Comment:
Can't task ID be retrieved once instead of doing a syscall all times?

Reply:
Yes, we could do that. There is even a gettid() implementation within rt-app already which currently is only declared for deadline sched class. So if I remove this DLSCHED restriction, store the thread id in thread_data_t and set it early in thread_body() then deadline and taskgroup support can both use them. What do you think?

```Author: Dietmar Eggemann <dietmar.eggemann@arm.com>
Date:   Thu Oct 18 19:22:28 2018 +0200

    Declare gettid() globally
    
    The function will be used by the taskgroup implementation. Remove the
    restriction that it is only declared in case the deadline class is
    supported.
    
    Signed-off-by: Dietmar Eggemann <dietmar.eggemann@arm.com>

diff --git a/src/rt-app_utils.h b/src/rt-app_utils.h
index 2570a91454db..c5cfb5bc0cf4 100644
--- a/src/rt-app_utils.h
+++ b/src/rt-app_utils.h
@@ -113,10 +113,10 @@ timespec_sub_to_ns(struct timespec *t1, struct timespec *t2);
 void
 log_timing(FILE *handler, timing_point_t *t);
 
-#ifdef DLSCHED
 pid_t
 gettid(void);
 
+#ifdef DLSCHED
 unsigned long long
 timespec_to_nsec(struct timespec *ts);
 #endif
```

```
Author: Dietmar Eggemann <dietmar.eggemann@arm.com>
Date:   Fri Oct 19 09:23:28 2018 +0200

    Fix: Retrieve thread id once
    
    Store the thread id (tid) inside thread_data_t so it can be used to set
    the thread into the different taskgroups without calling the kernel for
    the tid every time the thread changes taskgroups.
    
    The deadline scheduler class code in set_thread_priority() also use
    this new thread_data_t data member now directly.
    
    Signed-off-by: Dietmar Eggemann <dietmar.eggemann@arm.com>

diff --git a/src/rt-app.c b/src/rt-app.c
index c70497801827..9125f3601685 100644
--- a/src/rt-app.c
+++ b/src/rt-app.c
@@ -635,7 +635,6 @@ static void set_thread_priority(thread_data_t *data, sched_data_t *sched_data)
        policy_t policy;
 #ifdef DLSCHED
        struct sched_attr dl_params;
-       pid_t tid;
        unsigned int flags = 0;
 #endif
        int ret;
@@ -720,7 +719,6 @@ static void set_thread_priority(thread_data_t *data, sched_data_t *sched_data)
                                        policy_to_string(sched_data->policy), sched_data->period,
                                        sched_data->runtime, sched_data->deadline);
 
-                       tid = gettid();
                        dl_params.size = sizeof(struct sched_attr);
                        dl_params.sched_flags = 0;
                        dl_params.sched_policy = SCHED_DEADLINE;
@@ -729,7 +727,7 @@ static void set_thread_priority(thread_data_t *data, sched_data_t *sched_data)
                        dl_params.sched_deadline = sched_data->deadline;
                        dl_params.sched_period = sched_data->period;
 
-                       ret = sched_setattr(tid, &dl_params, flags);
+                       ret = sched_setattr(data->tid, &dl_params, flags);
                        if (ret != 0) {
                                log_critical("[%d] sched_setattr "
                                                "returned %d", data->ind, ret);
@@ -772,6 +770,9 @@ void *thread_body(void *arg)
                perror("pthread_setname_np thread name over 16 characters");
        }
 
+       /* Set thread id */
+       data->tid = gettid();
+
        /* Get the 1st phase's data */
        pdata = &data->phases[0];
 
diff --git a/src/rt-app_taskgroups.c b/src/rt-app_taskgroups.c
index 2bcfee1bf293..357969bb9875 100644
--- a/src/rt-app_taskgroups.c
+++ b/src/rt-app_taskgroups.c
@@ -86,7 +86,7 @@ static int cgroup_get_cpu_controller_mount_point(void)
        return ret;
 }
 
-static int cgroup_attach_task(char *name)
+static int cgroup_attach_task(char *name, pid_t tid)
 {
        char *path, *file;
        FILE *tasks;
@@ -113,7 +113,7 @@ static int cgroup_attach_task(char *name)
                goto error;
        }
 
-       if (fprintf(tasks, "%ld", syscall(SYS_gettid)) < 0) {
+       if (fprintf(tasks, "%ld", tid) < 0) {
                perror("fprintf");
                goto error;
        }
@@ -323,9 +323,9 @@ void set_taskgroup(thread_data_t *data, taskgroup_t *tg)
        if (!tg)
                return;
 
-       log_debug("[%d] setting taskgroup [%s]", data->ind, tg->name);
+       log_debug("[%d] set task [%s] taskgroup [%s]", data->ind, data->name, tg->name);
 
-       if (cgroup_attach_task(tg->name)) {
+       if (cgroup_attach_task(tg->name, data->tid)) {
                log_critical("cannot attach task to taskgroup [%s]", tg->name);
                exit(EXIT_FAILURE);
        }
diff --git a/src/rt-app_types.h b/src/rt-app_types.h
index c0b1eca5b811..301834a27e3a 100644
--- a/src/rt-app_types.h
+++ b/src/rt-app_types.h
@@ -178,6 +178,7 @@ typedef struct _phase_data_t {
 typedef struct _thread_data_t {
        int ind;
        char *name;
+       pid_t tid;
        int lock_pages;
        int duration;
        rtapp_resource_t **resources;
```

Reply:
Makes sense to me.